libbeat/processors/actions/alterFieldProcessor.go

```diff
@@ -0,0 +1,139 @@
+package actions
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/elastic/beats/v7/libbeat/beat"
+	"github.com/elastic/beats/v7/libbeat/processors"
+	conf "github.com/elastic/elastic-agent-libs/config"
+	"github.com/elastic/elastic-agent-libs/mapstr"
+	"github.com/pkg/errors"
+)
+
+// alterFieldFunc defines how fields must be processed
+type alterFieldFunc func(field string) string
+
+type alterFieldProcessor struct {
+	Fields        []string
+	IgnoreMissing bool
+	FailOnError   bool
+
+	processorName string
+	alterFunc     alterFieldFunc
+}
+
+// NewAlterFieldProcessor is an umbrella method for processing events based on provided fields. Such as converting event keys to uppercase/lowercase
+func NewAlterFieldProcessor(c *conf.C, processorName string, alterFunc alterFieldFunc) (processors.Processor, error) {
+	config := struct {
+		Fields        []string `config:"fields"`
+		IgnoreMissing bool     `config:"ignore_missing"`
+		FailOnError   bool     `config:"fail_on_error"`
+	}{
+		IgnoreMissing: false,
+		FailOnError:   true,
+	}
+
+	if err := c.Unpack(&config); err != nil {
+		return nil, fmt.Errorf("failed to unpack the %s fields configuration: %s", processorName, err)
+	}
+
+	// Skip mandatory fields
+	for _, readOnly := range processors.MandatoryExportedFields {
+		for i, field := range config.Fields {
+			if field == readOnly {
+				config.Fields = append(config.Fields[:i], config.Fields[i+1:]...)
+			}
+		}
+	}
+
+	return &alterFieldProcessor{
+		Fields:        config.Fields,
+		IgnoreMissing: config.IgnoreMissing,
+		FailOnError:   config.FailOnError,
+		processorName: processorName,
+		alterFunc:     alterFunc,
+	}, nil
+
+}
+
+func (a *alterFieldProcessor) String() string {
+	return fmt.Sprintf("%s fields=%+v", a.processorName, *a)
+}
+
+func (a *alterFieldProcessor) Run(event *beat.Event) (*beat.Event, error) {
+	var backup *beat.Event
+	if a.FailOnError {
+		backup = event.Clone()
+	}
+
+	for _, field := range a.Fields {
+		err := a.alter(event, field)
+		if err != nil {
+			if a.FailOnError {
+				event = backup
+				event.PutValue("error.message", err.Error())
+				return event, err
+			}
+		}
+	}
+
+	return event, nil
+}
+
+func (a *alterFieldProcessor) alter(event *beat.Event, field string) error {
+
+	// Get all matching keys
+	allMatchingKeys := getCaseInsensitiveKeys(event.Fields, field)
+	if allMatchingKeys == nil {
```

Comment:
It's always safer to check the length, it might be just an empty list, not `nil`.

```suggestion
	if len(allMatchingKeys) == 0 {
```